vaadin-cdi/src/main/java/com/vaadin/cdi/internal/AbstractVaadinContext.java

```diff
@@ -204,12 +204,12 @@ void dropSessionData(VaadinSessionDestroyEvent event) {
     private synchronized void dropUIData(SessionData sessionData, int uiId) {
         getLogger().fine("Dropping UI data for UI: " + uiId);
 
-        for (Entry<Contextual<?>, ContextualStorage> entry : new ArrayList<Entry<Contextual<?>, ContextualStorage>>(
-                sessionData.getStorageMap().entrySet())) {
-            Contextual<?> key = entry.getKey();
-            if (key instanceof UIContextual
-                    && ((UIContextual) key).getUiId() == uiId) {
-                destroy(entry.getKey());
+        Map<Contextual<?>, ContextualStorage> storageMap = sessionData.getStorageMap();
+        for (Contextual<?> contextual : new ArrayList<>(storageMap.keySet())) {
+            if (contextual instanceof UIContextual
+                    && ((UIContextual) contextual).getUiId() == uiId) {
+                final ContextualStorage storage = storageMap.remove(contextual);
```

Comment:
I've forgot to mention this pull request fixes a memory leak too.
ContextualStorage left in storageMap with all UIScoped beans inside. Except UI, because UI the one what is destroyed.
Current code does not call remove on storageMap until Session close.